src/backend/columnar/columnar_tableam.c

```diff
@@ -2579,3 +2598,325 @@ downgrade_columnar_storage(PG_FUNCTION_ARGS)
+	extensionsResultSet->econtext = GetPerTupleExprContext(estate);
+	extensionsResultSet->allowedModes = SFRM_Materialize;
+
+	fmgr_info(F_PG_AVAILABLE_EXTENSIONS, &flinfo);
+	InitFunctionCallInfoData(*fcinfo, &flinfo, 0, InvalidOid, NULL,
+							 (Node *) extensionsResultSet);
+
+	/* pg_available_extensions returns result set containing all available extensions */
+	(*pg_available_extensions)(fcinfo);
+
+	TupleTableSlot *tupleTableSlot = MakeSingleTupleTableSlotCompat(
+		extensionsResultSet->setDesc,
+		&TTSOpsMinimalTuple);
+	bool hasTuple = tuplestore_gettupleslot(extensionsResultSet->setResult, goForward,
+											doCopy,
+											tupleTableSlot);
+	while (hasTuple)
+	{
+		bool isNull = false;
+
+		Datum extensionNameDatum = slot_getattr(tupleTableSlot, 1, &isNull);
+		char *extensionName = NameStr(*DatumGetName(extensionNameDatum));
+		if (strcmp(extensionName, "citus") == 0)
+		{
+			Datum availableVersion = slot_getattr(tupleTableSlot, 2, &isNull);
+
+			/* we will cache the result of citus version to prevent catalog access */
+			/* MemoryContext oldMemoryContext = MemoryContextSwitchTo( */
+			/*	MetadataCacheMemoryContext_Columnar); */
+
+			availableExtensionVersion = text_to_cstring(DatumGetTextPP(availableVersion));
+
+			/* MemoryContextSwitchTo(oldMemoryContext); */
+
+			ExecClearTuple(tupleTableSlot);
+			ExecDropSingleTupleTableSlot(tupleTableSlot);
+
+			return availableExtensionVersion;
+		}
+
+		ExecClearTuple(tupleTableSlot);
+		hasTuple = tuplestore_gettupleslot(extensionsResultSet->setResult, goForward,
+										   doCopy, tupleTableSlot);
+	}
+
+	ExecDropSingleTupleTableSlot(tupleTableSlot);
+
+	ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					errmsg("citus extension is not found")));
+}
+
+
+/*
+ * InstalledExtensionVersion returns the Citus version in PostgreSQL pg_extension table.
+ */
+static char *
+InstalledExtensionVersion_Columnar(void)
+{
+	ScanKeyData entry[1];
+	char *installedExtensionVersion = NULL;
+
+	Relation relation = table_open(ExtensionRelationId, AccessShareLock);
+
+	ScanKeyInit(&entry[0], Anum_pg_extension_extname, BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum("citus"));
+
+	SysScanDesc scandesc = systable_beginscan(relation, ExtensionNameIndexId, true,
+											  NULL, 1, entry);
+
+	HeapTuple extensionTuple = systable_getnext(scandesc);
+
+	/* We assume that there can be at most one matching tuple */
+	if (HeapTupleIsValid(extensionTuple))
+	{
+		int extensionIndex = Anum_pg_extension_extversion;
+		TupleDesc tupleDescriptor = RelationGetDescr(relation);
+		bool isNull = false;
+
+		Datum installedVersion = heap_getattr(extensionTuple, extensionIndex,
+											  tupleDescriptor, &isNull);
+
+		if (isNull)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("citus extension version is null")));
+		}
+
+		/* we will cache the result of citus version to prevent catalog access */
```

Comment:
This had to be commented out but I'm not sure if it's needed or not

Reply:
Same as my other comment -- just make sure to free the result in the caller.